source/stun_serializer.c

```diff
@@ -0,0 +1,860 @@
+#define STUN_READ_UINT16    ( pCtx->readWriteFn.ReadU16Fn )
+#define STUN_READ_UINT32    ( pCtx->readWriteFn.ReadU32Fn )
+#define STUN_READ_UINT64    ( pCtx->readWriteFn.ReadU64Fn )
+
+static StunResult_t AddAttributeBuffer( StunContext_t * pCtx,
+                                        StunAttributeType_t attributeType,
+                                        const uint8_t * pAttributeValueBuffer,
+                                        uint16_t attributeValueBufferLength );
+
+static StunResult_t AddAttributeFlag( StunContext_t * pCtx,
+                                      StunAttributeType_t attributeType );
+
+static StunResult_t AddAttributeU32( StunContext_t * pCtx,
+                                     StunAttributeType_t attributeType,
+                                     uint32_t attributeValue );
+
+static StunResult_t CheckAndUpdateAttributeFlag( StunContext_t * pCtx,
+                                                 StunAttributeType_t attributeType );
+
+/*-----------------------------------------------------------*/
+
+static StunResult_t CheckAndUpdateAttributeFlag( StunContext_t * pCtx,
+                                                 StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    if( ( pCtx->attributeFlag & STUN_FLAG_FINGERPRINT_ATTRIBUTE ) != 0 )
+    {
+        /* No more attributes can be added after Fingerprint - it must  be
+         * the last attribute. */
+        result = STUN_RESULT_INVALID_ATTRIBUTE_ORDER;
+    }
+    else if( ( ( pCtx->attributeFlag & STUN_FLAG_INTEGRITY_ATTRIBUTE ) != 0 ) &&
+             ( attributeType != STUN_ATTRIBUTE_TYPE_FINGERPRINT ) )
+    {
+        /* No attribute other than fingerprint can be added after Integrity
+         * attribute. */
+        result = STUN_RESULT_INVALID_ATTRIBUTE_ORDER;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        /* Update flags. */
+        if( attributeType == STUN_ATTRIBUTE_TYPE_FINGERPRINT )
+        {
+            pCtx->attributeFlag |= STUN_FLAG_FINGERPRINT_ATTRIBUTE;
+        }
+        else if( attributeType == STUN_ATTRIBUTE_TYPE_MESSAGE_INTEGRITY )
+        {
+            pCtx->attributeFlag |= STUN_FLAG_INTEGRITY_ATTRIBUTE;
+        }
+    }
+
+    return result;
+}
+
+static StunResult_t AddAttributeBuffer( StunContext_t * pCtx,
+                                        StunAttributeType_t attributeType,
+                                        const uint8_t * pAttributeValueBuffer,
+                                        uint16_t attributeValueBufferLength )
+{
+    StunResult_t result = STUN_RESULT_OK;
+    uint16_t attributeValueLengthPadded = STUN_ALIGN_SIZE_TO_WORD( attributeValueBufferLength );
+
+    if( ( pCtx == NULL ) ||
+        ( pAttributeValueBuffer == NULL ) ||
+        ( attributeValueBufferLength == 0 ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( ( result == STUN_RESULT_OK ) &&
+        ( pCtx->pStart != NULL ) )
```

Comment:
Why is pCtx->pStart check needed here, since we are adding the check below as well?

Reply:
pCtx->pStart is NULL it means we are only calculating size, where as when it is not NULL we need to process accordingly.